### PR TITLE
[zero] add state dict for low level zero

### LINKED ISSUE #4154: [zero] override `state_dict` for low level zero
Since the fp32 params have been sharded evenly, the `state_dict` has been actually destroyed.

Therefore, we need to override the `state_dict` method to support load or save checkpoint.

### PR DESCRIPTION
**Pull request by Gy-Lu**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

closed #4154 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

Provide interface of low level zero's state_dict and load_state_dict.

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Comment by Gy-Lu** [comment]
The current implementation of `load_state_dict` would change the passing `state_dict`.
Not sure if it would introduce some bugs under some situation.

File: colossalai/zero/low_level/low_level_optim.py [hunk 1]
```diff
@@ -468,3 +469,66 @@ def no_sync(self):
...
+
+        Args:
+            state_dict (dict): A pytorch form state_dict, be aware that the state_dict would be changed
+        """
+        for param_idx, state in state_dict['state'].items():
+            for k, v in state.items():
+                if isinstance(v, torch.Tensor) and k != 'step':
+                    padding_size = (self._world_size - v.numel() % self._world_size) % self._world_size
+                    with torch.no_grad():
+                        v = v.clone().detach().flatten()
```
Comment (ver217) [inline colossalai/zero/low_level/low_level_optim.py:528]: Clone and detach after splitting can be more efficient.

Comment (Gy-Lu) [reply colossalai/zero/low_level/low_level_optim.py]: done

**Review by ver217 — APPROVED** [review]
